remotespark/sparkkernelbase.py

```diff
@@ -81,19 +60,74 @@ def do_execute(self, code, silent, store_history=True, user_expressions=None, al
-            self.execute_cell_for_user(code, True, False)
+            self._execute_cell_for_user(code, True, False)
             self.already_ran_once = False
 
-        return self.do_shutdown_ipykernel(restart)
+        return self._do_shutdown_ipykernel(restart)
+
+    def _initialize_magics(self, username, password, url):
+        connection_string = get_connection_string(url, username, password)
+
+        register_magics_code = "%load_ext remotespark"
+        self._execute_cell(register_magics_code, True, False, shutdown_if_error=True,
+                           log_if_error="Failed to load the Spark magics library.")
+        self.logger.debug("Loaded magics.")
+
+        add_endpoint_code = "%spark add {} {} {} skip".format(
+            self.client_name, self.session_language, connection_string)
+        self._execute_cell(add_endpoint_code, True, False, shutdown_if_error=True,
+                           log_if_error="Failed to create a Livy session.")
+        self.logger.debug("Added endpoint.")
 
-    def execute_cell_for_user(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False):
+        self.already_ran_once = True
+
+    def _get_configuration(self):
+        try:
+            username = get_configuration(self.username_conf_name)
+            password = get_configuration(self.password_conf_name)
+            url = get_configuration(self.url_conf_name)
+            return username, password, url
+        except KeyError:
+            message = "Please set configuration for '{}', '{}', '{} to initialize Kernel.".format(
+                self.username_conf_name, self.password_conf_name, self.url_conf_name)
+            self._abort_with_fatal_error(message)
+
+    def _execute_cell(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False,
+                      shutdown_if_error=False, log_if_error=None):
+        reply_content = self._execute_cell_for_user(code, silent, store_history, user_expressions, allow_stdin)
+
+        if shutdown_if_error and reply_content[u"status"] == u"error":
+            error_from_reply = reply_content[u"evalue"]
+            if log_if_error is not None:
+                message = "{}\nException details:\n\t\"{}\"".format(log_if_error, error_from_reply)
+                self._abort_with_fatal_error(message)
+
+        return reply_content
+
+    def _execute_cell_for_user(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False):
         return super(SparkKernelBase, self).do_execute(code, silent, store_history, user_expressions, allow_stdin)
 
-    def do_shutdown_ipykernel(self, restart):
+    def _do_shutdown_ipykernel(self, restart):
         return super(SparkKernelBase, self).do_shutdown(restart)
+
+    def _abort_with_fatal_error(self, message):
+        self._fatal_error = message
+
+        error = get_configuration(Constants.fatal_error_suggestion, self.fatal_error_suggestion).format(message)
```

Comment:
From a design perspective I'd say it's more sensible for default values to be handled by the get_configuration function (or, maybe better, a Configuration class).  Having default values for configurations be spread out through the entire codebase seems messy.